### DISCUSSION TITLE
Import-Module modulename -UseWindowsPowerShell won't remember global variables. Bug or intended?

### DISCUSSION THREAD

**Discussion by yllekz** [PowerShell/PowerShell#16390] [Q&A] [answered]
Hi folks, I'm trying to unravel a mystery. I don't want to submit this as an issue yet because honestly I'm not sure if it is one, but the results are confusing and inconsistent to me so I'm suspicious.

I've been trying to understand why PowerShell 7 seems to handle retaining global variables in inconsistent ways. It seems to come down to the ``-UseWindowsPowerShell`` switch when using ``import-module``.

I ran three simple tests to observe the inconsistency:

**The following results in global variables being retained in the console window:**
- Put this directly in PowerShell 7 Microsoft.PowerShell_profile.ps1:
- ``New-Variable -Scope Global -Force -Name 'glb_foo' -Value 'bar'``
- ``get-variable | where-object {$_.Name -match 'glb_'}``

**<For the next two tests below, I made a sample function for setting a global variable:>**

```
<#
.Synopsis
    Make a global variable in a function as a test to compare PowerShell 5.1/7 behavior differences
.DESCRIPTION
    Make a global variable in a function as a test to compare PowerShell 5.1/7 behavior differences
.EXAMPLE
    New-GlobalVarTest -GlbVarName 'glb_foo' -GlbVarVal 'bar'
.INPUTS
    Global Variable name, Global Variable value
.OUTPUTS
    A global variable
.NOTES
    Version: 1.0.0
.COMPONENT
    Testing only
.ROLE
    No idea
.FUNCTIONALITY
    Testing
#>
function New-GlobalVarTest
{
    [CmdletBinding(DefaultParameterSetName='Set 1',
                  SupportsShouldProcess=$true,
                  PositionalBinding=$false,
                  HelpUri = 'https://www.duckduckgo.com/',
                  ConfirmImpact='Medium')]
    [Alias()]
    [OutputType([String])]
    Param
    (
        #Variable Name
        [Parameter(Mandatory=$true,
                   ValueFromPipeline=$true,
                   ValueFromPipelineByPropertyName=$true,
                   ValueFromRemainingArguments=$false,
                   Position=0,
                   ParameterSetName='Set 1')]
        [ValidateNotNullOrEmpty()]
        [string]$GlbVarName,

        #Variable Value
        [Parameter(Mandatory=$true,
                   ValueFromRemainingArguments=$false,
                   Position=1,
                   ParameterSetName='Set 1')]
        [string]$GlbVarVal
    )

    Begin{}

    Process{
        if ($pscmdlet.ShouldProcess("$($glbVarName)", "Make global variable")){

            new-variable -Scope global -Force -Name $GlbVarName -Value $GlbVarVal

            $ReturnMe = (get-variable | where-object {$_.Name -match $GlbVarName})

        }
    }

    End{
        return $ReturnMe
    }
}
```

**The following results in global variables being retained in the console window:**
- PowerShell 7 Microsoft.PowerShell_profile.ps1 imports module without ``-UseWindowsPowerShell`` and calls ``New-GlobalVarTest -GlbVarName 'glb_foo' -GlbVarVal 'bar'``
- glb_ variables are loaded into the console window as expected

**The following results in NO global variables being retained in the console window:**
- PowerShell 7 Microsoft.PowerShell_profile.ps1 imports a module using ``-UseWindowsPowerShell`` and calls ``New-GlobalVarTest -GlbVarName 'glb_foo' -GlbVarVal 'bar'``
- Observe that no "GLB_" variables are loaded into the console window. This is the problem.

Is it expected behavior to not have global variables stick when calling a function that is part of a module imported with ``-UseWindowsPowerShell``?

**Comment by vexx32** [comment, marked as answer]
Importing a module with `-UseWindowsPowerShell` does a couple things:

1. Starts Windows PowerShell in the background
2. Loads the requested module in _that_ process
3. Opens communication pipes to that Windows PowerShell process to send/receive data
4. Creates a dummy module by the same name with a bunch of the same command names that will redirect command calls to the Windows PowerShell session.

As it's a completely separate process, things local to the PowerShell process will not be carried over to the Windows PowerShell process.

It's possible that future improvements could result in behaviour like this working to some extent, but it's by no means a simple scope of work, and it would never work for everything. Any data being sent to the Windows PowerShell process must be serialised/deserialised during the transfer, so there will always be cases that simply don't work correctly. IMO it's probably not worth the scope of work to get some of the things working, because there are a lot of use cases that simply won't work without a lot of additional work.

The `-UseWindowsPowerShell` switch is best used very selectively and only when strictly necessary. Likely the only really useful use-case is to load specifically Windows-only modules that don't yet work in PowerShell 7+, but even those use cases are not perfect, there are cases things won't work quite as expected.
